### ISSUE TITLE
[Package]: VSCode / VSCodium

### ISSUE THREAD

**Issue by dariotarantini** [termux/termux-packages#7641]
### Package description

VSCode is a powerful code editor available on ARM / ARM64. It can be installed under a proot enviroment, but it would be nice to have directly inside termux repo

### Home page URL

https://code.visualstudio.com

### Source code URL

https://github.com/VSCodium/vscodium

### Packaging policy acknowledgement

- [X] I have read and understand the [Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy).

**Comment by thunder-coding** [comment]
VS Code port would require first patching Electron to work under Termux. And patching Electron means we also need to patch Chromium. Both Chromium and Electron take a very high amount of time to build even on CI. While compiling nodejs, V8 takes more than an hour so wonder how much time Chromium will take. And even then if you manage to get Chromium, Electron and VS Code working on Termux. Keeping it maintained is not gonna be that easy. So I feel maintainers may not accept this package 🙂

And anyways did you try out [`code-server`](https://github.com/cdr/code-server), it allows you to use VS Code in your browser

Installation: 
```
apt install nodejs-lts yarn
yarn global add code-server
```

> Note: nodejs-lts is required instead of nodejs because code-server only supports 14.x of node for now.

**Comment by dariotarantini** [comment]
@thunder-coding yep, i know it. But i think that it will be awesome to port Electron on termux, thanks to all.app developed with it. I just hope that someone will start working on it. also, thanks, i already knew code-server but i will still use prooted distro with vscode

**Comment by Maxython** [comment]
The idea is cool but difficult.  I will try to compile `vscode` via `makepkg`.

**Comment by Maxython** [comment]
To compile `vscode` you need packages that are not in termux.  
![Screenshot_20211009-110918_Termux](https://user-images.githubusercontent.com/60917834/136650198-e35fbd07-856b-4d4d-8461-68c27c70d20c.jpg)
I'll take care of adding them to termux.

**Comment by thunder-coding** [comment]
> To compile vscode need packages that are not in termux.

Give here a list of them, and I will see them

**Comment by Maxython** [comment]
At the moment, here is the list.
- [X] keytar
- [ ] electron
  - [ ] playwright

Those that are not checked are not installed on termux.

**Comment by Maxython** [comment]
PS: `electron` is not installed due to a problem with `playwright`.

**Comment by thunder-coding** [comment]
> PS: `electron` is not installed due to a problem with `playwright`.

Can you open up a draft PR so that I can see what's in the log

**Comment by Maxython** [comment]
The result of the `playwright` compilation:
![Screenshot_20211009-160143_Termux](https://user-images.githubusercontent.com/60917834/136658945-6503f659-ef95-4e67-849b-488335ba0f6d.jpg)

**Comment by thunder-coding** [comment]
> The result of the `playwright` compilation:
> ![Screenshot_20211009-160143_Termux](https://user-images.githubusercontent.com/60917834/136658945-6503f659-ef95-4e67-849b-488335ba0f6d.jpg)
> 

Should be fixed by patching but first someone needs to add Chromium to x11-repo

**Comment by Maxython** [comment]
> > The result of the `playwright` compilation:
> > ![Screenshot_20211009-160143_Termux](https://user-images.githubusercontent.com/60917834/136658945-6503f659-ef95-4e67-849b-488335ba0f6d.jpg)
> 
> Should be fixed by patching but first someone needs to add Chromium to x11-repo

This is exactly what I am doing.

**Comment by Maxython** [comment]
The result of compiling `chromium` according to the advice from the developers:  
![Screenshot_20211010-232658_Termux](https://user-images.githubusercontent.com/60917834/136752214-8e1bc856-ccab-457f-940e-42313ac0e1d8.jpg)

**Comment by thunder-coding** [comment]
> The result of compiling `chromium` according to the advice from the developers:  
> ![Screenshot_20211010-232658_Termux](https://user-images.githubusercontent.com/60917834/136752214-8e1bc856-ccab-457f-940e-42313ac0e1d8.jpg)
> 

You need to fetch the source code using `depot_tools`' `gclient`  https://www.chromium.org/developers/how-tos/get-the-code

**Comment by Maxython** [comment]
Also `gn` is probably needed

**Comment by tathastu871** [comment]
use proot distro to install vscode. Andronix created distros have it

**Comment by Tsnow95** [comment]
> VS Code port would require first patching Electron to work under Termux. And patching Electron means we also need to patch Chromium. Both Chromium and Electron take a very high amount of time to build even on CI. While compiling nodejs, V8 takes more than an hour so wonder how much time Chromium will take. And even then if you manage to get Chromium, Electron and VS Code working on Termux. Keeping it maintained is not gonna be that easy. So I feel maintainers may not accept this package 🙂
> 
> And anyways did you try out [`code-server`](https://github.com/cdr/code-server), it allows you to use VS Code in your browser
> 
> Installation:
> 
> ```
> apt install nodejs-lts yarn
> yarn global add code-server
> ```
> 
> > Note: nodejs-lts is required instead of nodejs because code-server only supports 14.x of node for now.

error code-server@3.12.0: The engine "node" is incompatible with this module. Expected version "= 14". Got "17.3.0"

Do you know a way to install previous version or am I stuck outta luck?

**Comment by fanqianduan** [comment]
is there any update?

**Comment by ddkwork** [comment]
any info?

**Comment by licy183** [comment]
Added electron (just the launcher) in https://github.com/termux-user-repository/tur/commit/1a2c9b19c5d39fa7e721fbf32f86bd61505aa80b. Hope that will help.

**Comment by licy183** [comment]
Added in https://github.com/termux-user-repository/tur/commit/091aaf938540b9735f475884c9bd57851925baf1. It downloads pre-built electron binaries from https://github.com/termux-user-repository/electron-tur-builder.
I wonder if it is acceptable to add this to the main repository.

**Comment by PillFall** [comment]
I do not know if this would be off-topic, but with [vscode.dev](vscode.dev) is not easier to compile only [VSCode CLI](https://github.com/microsoft/vscode/tree/main/cli) (which is written in Rust) and connect to VSCode.dev via tunnel?

I tried for some time now, and the only package that show errors (some can be patched easily) is [`keyring`](https://github.com/hwchen/keyring-rs). However, I believe the remaining errors are shown because Termux does not have any `secrets-manager`, as far as I know.

**Comment by GitMensch** [comment]
just checking the state - what I wanted to do is to run the REH version of vscodium to connect from another machine - the standing issue is that the node modules seem to need a (cross?)compile for android...

**Comment by TomJo2000** [comment]
- Closed by termux-user-repository/tur#213

Can be moved to the termux/termux-packages repo after the build infrastructure for `chromium` and `electron` is ported.

Closing this issue in the meantime.
